src/DidCache.ts

```diff
@@ -1,16 +1,338 @@
+      batchFileHash: operation.batchFileHash,
+      type: operation.type
+    };
+
+    // If this is a duplicate of an earlier operation, we can
+    // ignore this operation. Note that we might have a previous
+    // operation with the same hash, but that previous operation
+    // need not be earlier in timestamp order - hence the check
+    // with lesser().
+    const prevOperation = this.opHashToInfo.get(opHash);
+    if (prevOperation !== undefined && lesser(prevOperation, opInfo)) {
+      return null;
+    }
+    // Update our mapping of operation hash to operation info overwriting
+    // previous info if it exists
+    this.opHashToInfo.set(opHash, opInfo);
+
+    // For operations that have a previous version, we need additional
+    // bookkeeping
+    if (operation.previousOperationHash) {
+      this.applyOpWithPrev(opHash, opInfo, operation.previousOperationHash);
+    }
+
+    return opHash;
+  }
+
+  public constructor (
+    private readonly cas: Cas,
+    private readonly didDocUpdate: DidDocumentUpdater,
+    private readonly didDocCreate: DidDocumentCreator) {
+
+  }
+
+  /**
+   * Rollback the state of the DidCache by removing all operations
+   * with transactionNumber greater than the provided parameter value.
+   * The intended use case for this method is to handle rollbacks
+   * in the blockchain.
+   *
+   * The current implementation is inefficient: It simply scans the two
+   * hashmaps storing the core Did state and removes all entries with
+   * a greater transaction number.  In future, the implementation should be optimized
+   * for the common case by keeping a sliding window of recent operations.
+   */
+  public rollback (transactionNumber: number) {
+
+    // Iterate over all nextVersion entries and remove all versions
+    // with "next" operation with transactionNumber greater than the provided
+    // parameter.
+    this.nextVersion.forEach((opHash, version, map) => {
+      const opInfo = this.opHashToInfo.get(opHash) as OperationInfo;
+      if (opInfo.transactionNumber > transactionNumber) {
+        map.delete(version);
+      }
+    });
+
+    // Iterate over all operations and remove those with with
+    // transactionNumber greater than the provided parameter.
+    this.opHashToInfo.forEach((opInfo, opHash, map) => {
+      if (opInfo.transactionNumber > transactionNumber) {
+        map.delete(opHash);
+      }
+    });
+  }
+
+  /**
+   * Resolve a did.
+   */
+  public async resolve (did: VersionId): Promise<DidDocument | null> {
+    const latestVersion = this.last(did);
+
+    // lastVersion === null implies we do not know about the did
+    if (latestVersion === null) {
+      return null;
+    }
+
+    return this.lookup(latestVersion);
+  }
+
+  /**
+   * Returns the Did document for a given version identifier.
+   */
+  public async lookup (versionId: VersionId): Promise<DidDocument | null> {
+    // Version id is also the operation hash that produces the document
```

Comment:
It is unclear to me where signature verification will be performed, we don't have to do them now, but can you add a TODO comment inline indicating where the code will be invoked?

Reply:
Added a TODO where signature verification could go.